README.rst
***********************************
EXT: Change front end user password
***********************************
Extension Key: ``changepassword``

Copyright 2011, Muhammed Alat, <info [mailat] alatsystems [domaindot] de>


============
Introduction
============

What does it do?
================
Changepassword is a simple frontend plugin to change the password
of front end users.
It supports changing plain text passwords, md5 passwords and
saltedpasswords password.


Screenshots
===========

.. image:: Documentation/changepassword.png
   :alt: Screenshot of changepassword


============
Installation
============
- Import ``changepassword`` from Extension repository and install .
- Create a page that will be displayed when fe users are logged in and
  want to change their passwort.
- Now you can create a page content where users can change their passwords.
  Here you may use the ``changepassword`` FE Plugin on this page to display
  the changepassword form.


=============
Configuration
=============
You can change the default template by selecting another from your
``fileadmin`` folder.
Just copy the original template from
``typo3conf/ext/changepassword/res/template.html`` to
``fileadmin/changepassword.html`` and change the layout.
Furthermore you habe to select a source Page where the front end users
are stored.
Don't forget to select the stored Page.
In the example below the page “Benutzer” with the ``id`` 3 is selected.

.. image:: Documentation/plugin-configuration.png
   :alt: Plugin configuration in the TYPO3 backend


=========
Reference
=========
To be updated.


=======
Samples
=======
If you have a subpart ``CHANGEPASSWORD`` you can load the plugin via the
following TypoScript configuration::

    page.10.subparts.CHANGEPASSWORD < plugin.tx_changepassword_pi1

You are also free to load it in any other part of your TypoScript
configurations.


==============
Known problems
==============
There are no known problems at the moment.
If you find any bugs, please report them.
You can report them due to the online form http://www.alatsystems.de/kontakt/.


==========
To-Do list
==========
- Insert a constant with the minimum length of the newpassword,
  now it is hardcoded with 6.


=========
Changelog
=========
- 2011-07-08: Bug Fixes: Thanks to Boris Glusmann from http://www.horschler.eu
- 2010-03-12: Update Documentation
- 2010-03-10: Initial release


ChangeLog
2012-06-04 Andreas Bouché <info@flagbit.de>

	* Released version 0.7.2
	* [+BUGFIX] Check if salted passwords are enabled in the FE (Thanks to Till Wimmer for submitting a patch.)

2012-04-17 Andreas Bouché <info@flagbit.de>

	* Released version 0.7.1
	* [+FEATURE] Added support for multiple fe_user storage pages (Thanks to Claus Fassing for submitting a patch.)

2012-03-30 Frederic Gaus <info@flagbit.de>

	* Released version 0.6.0
	* UPD caretaker of this extension
	* FIX some typos in locallang
	* ADD TYPO3 4.6 compatibility
	* ADD support for salted passwords other than md5
	* FIX some coding guideline violations
	* refactoring to a more object oriented approach

pi1/class.tx_changepassword_pi1.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Muhammed Alat <alat@bizlogix.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */

require_once(PATH_tslib.'class.tslib_pibase.php');


/**
 * Plugin 'Change Password' for the 'changepassword' extension.
 *
 * @author Muhammed Alat <alat@bizlogix.de>
 * @author Frederic Gaus <gaus@flagbit.de>
 * @package TYPO3
 * @subpackage tx_changepassword
 */
class tx_changepassword_pi1 extends tslib_pibase {
	public $prefixId = 'tx_changepassword_pi1'; // Same as class name
	public $scriptRelPath = 'pi1/class.tx_changepassword_pi1.php'; // Path to this script relative to the extension dir.
	public $extKey = 'changepassword'; // The extension key.


	private function init($conf) {
		$this->conf = $conf;
		$this->pi_setPiVarDefaults();
		$this->pi_loadLL();
		$this->pi_initPIflexForm(); // Init FlexForm configuration for plugin

		// Get the logged in user id or return false
		$this->config['userid'] = $GLOBALS['TSFE']->loginUser ? $GLOBALS['TSFE']->fe_user->user['uid'] : false;

		// Getting the pid list via the flexform
		$pidList = $this->cObj->data['pages'];
		$this->conf['pidList'] = $pidList ?
			$pidList :
			$GLOBALS['TSFE']->id;

		// Template code
		$this->config['templateFile'] = $this->cObj->fileResource($this->pi_getFFvalue($this->cObj->data['pi_flexform'], 'templateFile', 'sDEF') ? $this->pi_getFFvalue($this->cObj->data['pi_flexform'], 'templateFile', 'sDEF')  : t3lib_extMgm::siteRelPath($this->extKey) . 'res/template.html');
	}


	/**
	 * The main method of the PlugIn
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	public function main($content, $conf) {
		$this->pi_USER_INT_obj = 1;
		$this->init($conf);
		if(!$this->config['userid']) {
			return 'User is not logged in.';
		}
		$content = '';

		// Get Part of the Template
		$template = $this->cObj->getSubpart($this->config['templateFile'], '###TEMPLATE###');
		$subTemplateErrOldPassword = $GLOBALS['TSFE']->cObj->getSubpart($template, '###SUBPART_ERR_OLDPASSWORD###');
		$subTemplateErrNewPassword = $GLOBALS['TSFE']->cObj->getSubpart($template, '###SUBPART_ERR_NEWPASSWORD###');
		$subTemplateErrRePassword = $GLOBALS['TSFE']->cObj->getSubpart($template, '###SUBPART_ERR_REPASSWORD###');
		$subpartArray = array(
			'###SUBPART_ERR_OLDPASSWORD###' => '',
			'###SUBPART_ERR_NEWPASSWORD###' => '',
			'###SUBPART_ERR_REPASSWORD###' => '',
		);

		$changePassword = false;
		if(isset($this->piVars['save'])) {
			$changePassword = true;
			if(! $this->isOldPasswordCorrect()) {
				$changePassword = false;
				$subMarkerArrayErrOldPassword = array();
				$subMarkerArrayErrOldPassword['###ERR_OLDPASSWORD###'] = $this->pi_getLL('wrongOldPassword');
				$subpartArray['###SUBPART_ERR_OLDPASSWORD###'] .= $this->cObj->substituteMarkerArrayCached($subTemplateErrOldPassword, $subMarkerArrayErrOldPassword, array(), array());
			} else if(! $this->isNewPasswordLongEnough()) {
				$changePassword = false;
				$subMarkerArrayErrNewPassword = array();
				$subMarkerArrayErrNewPassword['###ERR_NEWPASSWORD###'] = $this->pi_getLL('toshortpassword');
				$subpartArray['###SUBPART_ERR_NEWPASSWORD###'] .= $this->cObj->substituteMarkerArrayCached($subTemplateErrNewPassword, $subMarkerArrayErrNewPassword, array(), array());
			} else if(! $this->isNewPasswortEqual()) {
				$changePassword = false;
				$subMarkerArrayErrRePassword = array();
				$subMarkerArrayErrRePassword['###ERR_REPASSWORD###'] = $this->pi_getLL('twodifferentnewpasswords');
				$subpartArray['###SUBPART_ERR_REPASSWORD###'] .= $this->cObj->substituteMarkerArrayCached($subTemplateErrRePassword, $subMarkerArrayErrRePassword, array(), array());
			}
			if($changePassword) {
				$this->savePassword();
			}
		}

		$markerArray=array();
		if($changePassword) {
			$template = $this->cObj->getSubpart($this->config['templateFile'], '###TEMPLATE_SUCCESS###');
			$markerArray['###HEADER###'] = $this->pi_getLL('passwordsaved');
			$markerArray['###MESSAGE###'] = $this->pi_getLL('passwordsaved_message');
		} else {
			$markerArray['###FORM_ACTION###'] = $this->pi_getPageLink($GLOBALS['TSFE']->id);
			$markerArray['###LEGEND###'] =  $this->pi_getLL('legend');
			$markerArray['###OLDPASSWORD###'] =  $this->pi_getLL('oldpassword');
			$markerArray['###NEWPASSWORD###'] =  $this->pi_getLL('newpassword');
			$markerArray['###REPASSWORD###'] =  $this->pi_getLL('repassword');
			$markerArray['###SUBMIT###'] =  $this->pi_getLL('submit');
		}

		// build template
		$content .= $this->cObj->substituteMarkerArrayCached($template, $markerArray, $subpartArray,array());

		return $this->pi_wrapInBaseClass($content);
	}


	// Save the new password
	private function savePassword() {
		$updatePassword = array('password' => '', 'tstamp' => time());
		if (t3lib_extMgm::isLoaded('saltedpasswords') && tx_saltedpasswords_div::isUsageEnabled('FE')) {
			// EXT: saltedpassword
			$instanceSalted = tx_saltedpasswords_salts_factory::getSaltingInstance();
			$updatePassword['password'] = $instanceSalted->getHashedPassword($this->piVars['newpassword']);
		} else if (t3lib_extMgm::isLoaded('kb_md5fepw')) {
			// EXT: kb_md5fepw
			$updatePassword['password'] = md5($this->piVars['newpassword']);
		} else {
			// Plain Text
			$updatePassword['password'] = $this->piVars['newpassword'];
		}
		// Save new Password in the db
		$updateQueryStatus = $GLOBALS['TYPO3_DB']->exec_UPDATEquery (
			'fe_users',
			'uid = ' . $this->config['userid'] . ' AND pid IN (' . $this->conf['pidList'] . ')',
			$updatePassword
		);
		return $updateQueryStatus;
	}


	// Check the old password
	private function isOldPasswordCorrect() {
		// Check old password
		$password = $GLOBALS['TYPO3_DB']->exec_SELECTgetSingleRow (
			'password', // SELECT
			'fe_users', // FROM
			'uid = ' . $this->config['userid'] .' AND pid IN (' . $this->conf['pidList'] . ')' //WHERE
		);
		$password = current($password);
		if (t3lib_extMgm::isLoaded('saltedpasswords') && tx_saltedpasswords_div::isUsageEnabled('FE')) {
			$instanceSalted = tx_saltedpasswords_salts_factory::getSaltingInstance();
		}
		if ($instanceSalted && $instanceSalted->isValidSaltedPW($password)) {
			if (! $instanceSalted->checkPassword($this->piVars['oldpassword'], $password)) {
				return false;
			}
		} else if (t3lib_extMgm::isLoaded('kb_md5fepw')) {
			if(strcmp(md5($this->piVars['oldpassword']),$password)!=0) {
				return false;
			}
		} else {
			if(strcmp($this->piVars['oldpassword'],$password)!=0) {
				return false;
			}
		}
		return true;
	}


	// Check the length of the new password
	private function isNewPasswordLongEnough() {
		if( strlen($this->piVars['newpassword']) < 6 ) {
				return false;
		}
		return true;
	}


	// Check if the new password was enter two times and is equal
	private function isNewPasswortEqual() {
		if(strcmp($this->piVars['newpassword'],$this->piVars['repassword'])!=0) {
			return false;
		}
		return true;
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/changepassword/pi1/class.tx_changepassword_pi1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/changepassword/pi1/class.tx_changepassword_pi1.php']);
}

?>


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "changepassword".
#
# Auto generated 04-06-2012 15:23
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Change front end user password',
	'description' => 'Change the password of the front end user in the frontend. Easy template system to configure the layout. User have to enter his old password an twice the new one. Support for salted passwords.',
	'category' => 'fe',
	'shy' => 0,
	'version' => '0.7.2',
	'dependencies' => 'felogin',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'alpha',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Frederic Gaus',
	'author_email' => 'info@flagbit.de',
	'author_company' => 'Flagbit GmbH & Co. KG',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'felogin' => '',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:15:{s:9:"ChangeLog";s:4:"5957";s:12:"ext_icon.gif";s:4:"2359";s:17:"ext_localconf.php";s:4:"84d5";s:14:"ext_tables.php";s:4:"ab7d";s:15:"flexform_ds.xml";s:4:"34c7";s:16:"locallang_db.xml";s:4:"c113";s:14:"doc/manual.sxw";s:4:"cd17";s:19:"doc/wizard_form.dat";s:4:"dc9f";s:20:"doc/wizard_form.html";s:4:"c1dd";s:35:"pi1/class.tx_changepassword_pi1.php";s:4:"2caf";s:17:"pi1/locallang.xml";s:4:"f170";s:13:"res/style.css";s:4:"c053";s:17:"res/template.html";s:4:"ef53";s:35:"static/changepassword/constants.txt";s:4:"d41d";s:31:"static/changepassword/setup.txt";s:4:"d41d";}',
	'suggests' => array(
	),
);

?>